src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,920 @@
+                if (!queryExpressionProcessingInfo.ContainsIdentifier(forEachStatement.Identifier))
+                {
+                    documentUpdateInfo = ConvertIfInToForeachWithExtraVariableDeclaration(forEachStatement, queryExpressionProcessingInfo);
+                    return true;
+                }
+                else
+                {
+                    // The last select expression in the query returns this identifier:
+                    // foreach(var thisIdentifier in from ....... select thisIdentifier)
+                    // if thisIdentifier in foreach is var, it is OK
+                    // if a type is specified for thisIdentifier in forEach, check that the type is the same as in the select expression
+                    // foreach(MyType thisIdentifier in from ....... from MyType thisIdentifier ... select thisIdentifier)
+
+                    // The last clause in query stack must be SelectClauseSyntax.
+                    var lastSelectExpression = ((SelectClauseSyntax)queryExpressionProcessingInfo.Stack.Peek()).Expression;
+                    if (lastSelectExpression is IdentifierNameSyntax identifierName)
+                    {
+                        if (forEachStatement.Identifier.ValueText == identifierName.Identifier.ValueText)
+                        {
+                            var forEachStatementTypeSymbol = _semanticModel.GetTypeInfo(forEachStatement.Type, _cancellationToken).Type;
+                            var lastSelectExpressionTypeSymbol = _semanticModel.GetOperation(lastSelectExpression, _cancellationToken).Type;
+                            if (SymbolEquivalenceComparer.Instance.Equals(forEachStatementTypeSymbol, lastSelectExpressionTypeSymbol))
+                            {
+                                documentUpdateInfo = ConvertIfInToForeachWithoutExtraVariableDeclaration(forEachStatement, queryExpressionProcessingInfo);
+                                return true;
+                            }
+                        }
+                    }
+                }
+
+                // in all other cases try to replace with a local function - this is called above.
+                documentUpdateInfo = default;
+                return false;
+            }
+
+            private DocumentUpdateInfo ConvertIfInToForeachWithExtraVariableDeclaration(
+                ForEachStatementSyntax forEachStatement,
+                QueryExpressionProcessingInfo queryExpressionProcessingInfo)
+            {
+                // before foreach(var x in from ... a) { dosomething(x); }
+                // after 
+                // foreach (var a in ...)
+                // ...
+                // {
+                //      var x = a;
+                //      dosomething(x); 
+                //  }
+                var statements = GenerateStatements(
+                    expression => AddToBlockTop(SyntaxFactory.LocalDeclarationStatement(
+                    SyntaxFactory.VariableDeclaration(
```

Comment:
>SyntaxFactory.VariableDeclaration( [](start = 20, length = 34)

Alignment like this make it very hard to make sense what the code is doing. This is inside a lambda, why is it aligned with lambda's parameter?